Create explainer.md for rendering model (#63)
WICG/shared-element-transitions: explainer.md

This feature provides developers with the same capability on the web, irrespective of whether the scenes in a transition are rendered across Documents.

# Use-Cases
A visual demo of the transition patterns targeted by this feature are [here](https://material.io/design/motion/the-motion-system.html#transition-patterns). The following is a summary of the semantics of these transition patterns :

* Root Transitions : The full page content animates between 2 web pages with an optional static UI element on top. [Shared axis](https://material.io/design/motion/the-motion-system.html#shared-axis) shows an example.
* Shared Element Transitions : A persistent UI element morphs into another (which could be a UI element on the next page or the whole page) changing its shape and content. [Container transform](https://material.io/design/motion/the-motion-system.html#container-transform) shows an example.
* Entry/Exit Transitions : A UI element animates as it exits or enters the screen. This [issue](https://github.com/WICG/shared-element-transitions/issues/37) shows an example.

These transitions should be feasible in SPAs (Single Page Apps) and MPAs (Multi Page Apps).

# Design

> Let's take the example below which shows how the API can be used by a developer to animate the background and a shared element on a same origin navigation (MPA).

Reviewer: "The SPA equivalent of this case is one where the old document is mutated into the new document via DOM APIs."

> ### Old Document

Reviewer: ping?

```
<html>
<head>
  <style>
    body {
      background-color: blue;
    }
    .animated {
      width: 100px;
      height: 100px;

>       background-color: red;

Reviewer: Put position: relative; top: 300px so it's nonzero and different than below.

Author: Done.

    }
  </style>
</head>
<body>

>   <div class="animated" shared-id="shared-header" id="header"></div>

Author: Done.

</body>
</html>
```

### New Document
```
<html>
<head>
  <style>
    body {
      background-color: lightblue;
    }
    .animated {
      position: relative;
      top: 100px;
      width: 200px;
      height: 200px;
      background-color: red;
    }
    
    @keyframes fadeIn {
      0% {opacity: 0;}
      100% {opacity: 1;}
    }
    
    @keyframes fadeOut {
      0% {opacity: 1;}
      100% {opacity: 0;}
    }
    
    ::shared-new(#root), ::shared-new(#shared-header) {
      animation: fadeIn 1s;
    }

    ::shared-old(#root), ::shared-old(#shared-header) {
      animation: fadeOut 1s;
    }
    
    ::shared-container(#shared-header) {
      transition: all 1s ease-in;
    }
  </style>
</head>
<body>
  <div class="animated" shared-id="shared-header" id="header"></div>
</body>
</html>
```

The steps taken by the browser during the transition are as follows.

> 1. When a navigation is initiated on the old Document, create the following pseudo elements in the top layer :

Reviewer: ```suggestion
1. When a navigation is initiated from the old Document to the new Document, create the following pseudo elements in the top layer:
```

>     a. A container and child replaced element for each element with shared-id attribute. These are identified via ::shared-container(#shared-id) and ::shared-old(#shared-id) respectively.

Reviewer: ```suggestion
    a. A container and child replaced element for each element with the shared-id attribute. These are identified via ::shared-container(#shared-id) and ::shared-old(#shared-id) respectively.
```

>     a. A container and child replaced element for each element with shared-id attribute. These are identified via ::shared-container(#shared-id) and ::shared-old(#shared-id) respectively.

Reviewer 2: I'd verify the formatting, since in the rendered view these lines appear in the same line as above, which makes it look very confusing.

Also, do we need to explicitly call out that these pseudo would not have an associated ::backdrop (from https://fullscreen.spec.whatwg.org/#css-pe-backdrop), I'm guessing it's fine since pseudo elements can't have pseudo elements AFAIK.

Author: You're right that its probably obvious that a pseudo element won't have a pseudo element of its own but I added a footnote to mention this.

>     b. A replaced element for the root/html element identified via ::shared-old(#root).

Reviewer: 1.c The box hierarchy in the top layer is...

top layer stacking context:
  root
  container
    -> old

>     b. A replaced element for the root/html element identified via ::shared-old(#root).

Reviewer 2: a replaced pseudo-element or an element?

Author: I thought it was clear because the point says "create the following pseudo elements"?

Reviewer: +1 to 'replaced pseudo-element' to be extra clear.

Author: Ok. I switched it to "container pseudo-element and child replaced pseudo-element" to also make it clearer that only the child is a replaced element.

2. Apply the following UA stylesheet to the pseudo elements on the old page :
```

> ::shared-old(#root), ::shared-container(#shared-header) {

Reviewer: width height 100%?

  position: fixed;
  top: 0px;
  left: 0px;
}

::shared-container(#shared-header) {

>   // Container sized to the element's border-box size.

Reviewer: ```suggestion
  // This size is is chosen exactly according to the #shared-header element's border box dimensions after layout.
```

  width: 100px;
  height: 100px;
  

>   // A transform mapping the element to its quad in viewport space.

Reviewer: With my suggestion above, this should be:

transform: translate(0px, 308px)

Author: Done.

  transform: translate(0px, 0px);
}

::shared-old(#shared-header) {
  position: absolute;
  width: 100%;
  height: 100%;
  content: element(#header);
}

::shared-old(#root) {
  width: 100vw;
  height: 100vh;
  // The output of element() function on the root element.
  content: element(html);
}

> ```

Reviewer: #shared-header {
  opacity: 0
}

Reviewer: Alternative "element() doesn't recurse into shared element descendants"

Author: Another reason I was avoiding saying we're overriding the opacity to 0 on shared elements is because this will be visible to script which can't be done in cross-origin.

> 3. Save the output of the element() function for each pseudo element referenced above with the computed size and transform applied to container elements. Then navigate to the new page leaving the last rendered pixels of the old page on screen.

Reviewer: Split second sentence into another step.

> 4. When the new page loads, suppress rendering until resources required for first render have been fetched. This state is currently driven by browser heuristics but the eventual goal is to give deterministic control to developers using a [renderblocking](https://github.com/whatwg/html/issues/7131) attribute.

Reviewer: Remove second sentence.

> 5. Once the page is ready for first render, create the following pseudo elements in the top layer. This is done until the end of rAF callbacks on the page's first rendering lifecycle update[^1] :

Reviewer: I propose that we the top layer pseudo elements are "kept in sync with the corresponding shared elements in the DOM until the shared element transition animation completes in step 7."

Author: I like it! This also gives the page the ability to start animations using the old content sooner (if they already know the size/location for the new elements for instance) while the new content is streaming in.

    a. A container and child replaced element for each shared element on the old page using state saved in step 3.

>     b. A container and child replaced element for each element with shared-id attribute on the new page. Only the child element is created if a container with the matching id was already created in step a. above.

Reviewer: Rephrase second sentence to something like "the container is reused if already present"

Author: Done.

>     c. A replaced element for the root/html element identified via ::shared-old(#root).

Reviewer 2: Is root a reserved string in this case? Is that worth calling out? I'm not sure what's the intended behavior of 
`<div shared-id=root></div>`

Author: Added a comment to clarify that root is a reserved keyword.

6. Apply the following UA stylesheet to the pseudo elements on the new page.
```
::shared-old(#root), ::shared-new(#root), ::shared-container(#shared-header) {
  position: fixed;
  top: 0px;
  left: 0px;
}

::shared-old(#root), ::shared-new(#root) {
  width: 100vw;
  height: 100vh;
  content: element(html);
}
::shared-old(#root) {
  // This is the saved output referenced in step 3.
  content: cached-element(html);
}
::shared-new(#root) {
  content: element(html);
}

::shared-old(#shared-header), ::shared-new(#shared-header) {
  position: absolute;
  width: 100%;
  height: 100%;
  
  content: element(#header);
}
::shared-old(#shared-header) {
  // This is the saved output referenced in step 3.
  content: cached-element(#header);
}
::shared-new(#shared-header) {
  content: element(#header);
}

// Update the container's size and transform to the shared element on the new page.

> ::shared-container(#shared-header) {

Author: I added the bits for isolation to get a stacking context and mix-blend-mode.

> ::shared-container(#shared-header) {

Reviewer 2: You might want to move this block above `::shared-old(#shared-header)` since percent width/height depend on this width/height so it might read easier if you have this first. Up to you

Author: Good idea. Done.

  width: 200px;
  height: 200px;
  transform: translate(0px, 100px);

> }

Reviewer: Add an animation here.

Reviewer: ...0.5s

Author: Done. At 0.25s since that's the default we have right now.

```

> 7. Once no pseudo element has an active animation, remove them from the top layer.

Reviewer: "The animation is complete when the animation specified in step 6 completes"

Author: That makes it sound like we're saying the transition finishes when the UA default animation has completed but this needs to wait for developer configured animations as well. I liked the idea of letting developers specify a time or perhaps a script API (transitionFinished) if they're doing a rAF driven animation. Let's leave this open for now. I'll file a Github issue to discuss and we can update based on the conclusion there.

> 7. Once no pseudo element has an active animation, remove them from the top layer.

Reviewer 2: Does this preclude the effect of "element doesn't move while the page underneath it shifts". Or is this an all or nothing?

Author: It's all or nothing. As long as any element has an animation, all top layer elements are alive. I'm definitely open to better alternatives here. Was planning to file an issue to discuss this and update the explainer.

> 

Reviewer: Add a sub-section with a demo for this exactly example based on firefox stuff.

Author: You mean a link to a jsbin like this one : https://jsbin.com/xodipakazu/edit?html,output. Except for this example.

Reviewer: yeah

> ## Live Animatable Properties

Reviewer: ```suggestion
##  Animating Box Decoration CSS Properties
```

Author: Done.

> A common capability desirable during transitions is to interpolate styles like border-radius that change the element's shape. The [container transform](https://material.io/design/motion/the-motion-system.html#container-transform) examples show a visual demo of that. Painting properties like the element's border within its image when using the element() function makes this difficult. The following example[^2] shows how developers can control which styles are captured in the element() function vs being applied to the pseudo container on the old page.

Reviewer: Makes it difficult, or results in low-quality rendering? If the latter, show an example jsbin.

Nit: line wrap at < 80 chars.

Author: Animations like this : https://jsbin.com/xodipakazu/edit?html,output (open in Firefox) are infeasible unless the box decorations are interpolated on the pseudo elements. So at least that part is necessary.

Leaving them in the snapshot is the part as well would result in low-quality rendering IMO.

Reviewer: sure, I'm just suggesting you add a link to a demo and clarify how much is difficulty of doing something at all vs high quality

```
<html>
<head>
  <style>
    .animated {
      width:100px;
      height:100px;
      border: 10px solid black;
      border-radius: 10% 10%;
      box-shadow: 0px 0px 10px;
      

>       &:transition {

Reviewer 2: Is this SASS? I think this should be pure CSS

Author: Was trying to do it without copying the entire declaration. I couldn't find a way to do that with pure CSS.

        /* Retain the border to ensure it is painted transparent but the box size is unchanged. */
        border: 10px solid transparent;
        border-radius: none;
        box-shadow: none;
      }
    }
    
    ::shared-container(#shared-header) {
      border: 10px solid black;
      border-radius: 10% 10%;
      box-shadow: 0px 0px 10px;
    }
    
    ::shared-old(#shared-header) {
      top: -10px;
      left: -10px;
    }
  </style>
</head>
<body>
  <div class="animated" shared-id="shared-header">hello world</div>
</body>
</html>
```

> This capability requires the following changes to the steps mentioned in the design above :

Reviewer: You should spell out exactly what happens, this paragraph is too vague.

Reviewer: Also, I think you are proposing that the developer can do it themselves, with the aid of the pseudo-class. Is that correct?

Author: I was tempted to merge it with the original example but keeping it seperate helps to show the additional steps required only for this functionality. Let me know what you think.

> * A new pseudo class ("transition") is introduced which is activated when an element is being displayed using pseudo elements in step 1 and 5.

Reviewer 2: By activated you mean it applies when the element is being displayed using pseudo elements and doesn't apply otherwise?

Also, I'll be honest I had to keep looking back at what steps 1 and 5 and 3 (below) are, it might be nicer to reference those steps but also briefly explain what they are again

Author: Added a comment to reiterate what the step was doing.

* When saving state in step 3, the container element's complete style is preserved instead of only the computed size and transform.

## Modifications to element()
The following changes will be made to the element() spec as a part of this proposal :

> * The target element must have paint containment (contain:paint) to ensure the element is the containing block for all positioned descendents and generates a stacking context.

Reviewer 2: I don't think you mentioned what a "target element" is

Author: Added a comment.

* The target element must disallow fragmentation (break-inside:avoid).

> * A new cached-element() function is introduced to refer to the saved output of the element() function in step 3) of the design.

Reviewer: Also: "nested shared elements are omitted from the output of element()"

Author: Done. And the special cases for the root element.

# API Examples
## SPA
TODO: Add example

## MPA Script Based
TODO: Add example

# Alternatives Considered
## Heirarchical Properties

> This proposal disallows a shared element to be nested inside another shared element. The restriction avoids the need to preserve the hierarchy of the shared elements and associated properties (transform, clip, effects inherited by descendents) when creating pseudo elements. This is a consideration for future iterations of the feature.

Reviewer: It does not allow nesting? I thought we were allowing that? It's the pseudos that don't nest in the top layer, not the shared elements right?

Author: There was a discussion about this in the context of how we'll add the heirarchical aspect incrementally later. Disallowing it for now makes it easy to avoid backward compatibility issues if we change the behaviour to retain the hierarchy in pseudo elements then.

## Container/Child Split

> One consideration was to render each shared element using a replaced element directly instead of creating a container element. The motivation behind this split is to provide a stacking context to cross-fade the content of old and new shared elements. This was necessary to ensure blending identical pixels is a no-op using [plus-lighter](https://drafts.fxtf.org/compositing/#porterduffcompositingoperators_plus_lighter) blending. While same-origin transitions could avoid this, it enables future extensibility for cross-origin transitions where cross-fading identical images would be common.

Reviewer: What do you mean by "avoid this"? Do you mean "work around this"?

Author: Yes. Done.

> ## Natively Supported Live Animatable Properties

Reviewer: ```suggestion
## Natively Supporting Animating Box Decoration CSS Properties
```

Author: Done.

> An alternate approach to the setup described in [Live Animatable Properties](#live-animatable-properties) is to support this natively in the browser by introducing a new content-element() function. This function would behave similar to the element() function except skipping the following properties when painting the element: box decorations and visual effects which generate a stacking context. The image will also be sized to the element's content-box (as opposed to the border-box used by the element() function). The motivation for supporting this natively would be to make these properties animatable instead of requiring developer side changes.

Reviewer: ```suggestion
An alternate approach to the setup described in [Animating Box Decoration CSS Properties](#live-animatable-properties) is to support this natively in the browser by introducing a new content-element() function. This function would behave similar to the element() function except skipping the following properties when painting the element: box decorations and visual effects which generate a stacking context. The image will also be sized to the element's content-box (as opposed to the border-box used by the element() function). The motivation for supporting this natively would be to make these properties animatable instead of requiring developer side changes.
```

Author: Done.

This feature provides developers with the same capability on the web, irrespective of whether the scenes in a transition are rendered across Documents.

# Use-Cases
A visual demo of the transition patterns targeted by this feature are [here](https://material.io/design/motion/the-motion-system.html#transition-patterns). The following is a summary of the semantics of these transition patterns :

* Root Transitions : The full page content animates between 2 web pages with an optional static UI element on top. [Shared axis](https://material.io/design/motion/the-motion-system.html#shared-axis) shows an example.
* Shared Element Transitions : A persistent UI element morphs into another (which could be a UI element on the next page or the whole page) changing its shape and content. [Container transform](https://material.io/design/motion/the-motion-system.html#container-transform) shows an example.
* Entry/Exit Transitions : A UI element animates as it exits or enters the screen. This [issue](https://github.com/WICG/shared-element-transitions/issues/37) shows an example.

These transitions should be feasible in SPAs (Single Page Apps) and MPAs (Multi Page Apps).

> # Glossary

Reviewer: Rather than a long list of glossary and contents at the top, let's go straight to an example. I propose you give an example of navigating from one page that is a solid color with a single rectangle shared element on it, to a page that is some other solid color with the same shared element at a different position.

Old page:
<html style="background: blue">
  <div sharedId="header" style="width: 50px; height: 50px; background: green; position: relative"></div>
</html>

New page:
  <div sharedId="header" style="width: 50px; height: 50px; background: green; position: relative; top: 300px"></div>
</html>

The API would be something like:

Old document:
addEventListener("navigate", (event) => {
  document.offerTransition({sharedElements: ["header"]});
}

New one:
<html transition="fadeIn" transitionDuration=0.1s>
...

Then you can proceed directly to the steps:

1. When offerTransition is called:
1a. Save the output of element(sharedId) for each shared id offered.
1b. Save the output of element(html).
2. Navigate to the new page, while leaving the last rendered pixels of the old page on the screen.
3. If the new page has a "transition" attribute on the HTML element, suppress rendering until DOMContentLoaded is completed.
4. After DOMContentLoaded has fired, start an animation as follows:
4a. Create a container element named and its child in the top layer for each shared element |name| in the current document. The children are accessible via a pseudo-element selector ::new-shared-element(|name|)
4b. Create a container element and child for each shared element in the old document, and identify via ::old-shared-element(|name|)
4b. Apply a UA style sheet with ... in it. (element() function, opacity:0 on shared elements in the page, etc
4c. Render as normal
5. At the completion of the animation time specified in the transitionDuration attribute, remove all pseudo-elements from the top layer.

Then say that the following changes to the element() spec would be made:

1. containment
2. no breaking
3. no visual effects
4. element(thing in old document) refers to a snapshot from 1a/1b

If you say all this in the language of the CSSWG specs, you don't need a lot of the explanatory text about what "painted content" means.

Author: One more thing, the no visual effects modification seems difficult to explain in a way that would be backwards compatible. And it also opens the discussion for why only those styles and not box decorations. I feel like if we're suggesting that, we might as well be discussing the "Natively Supported Live Animatable Properties" alternative.

## Rendering Transition DOM
The DOM representation referenced above is rendered and animated by creating pseudo elements with the [content](https://drafts.csswg.org/css-content/#content-property) property set to a shared element's painted content, the [transform](https://drafts.csswg.org/css-transforms-1/#propdef-transform) set to the viewport space transform and width/height set to the element's border-box size. These pseudo elements are drawn on top of the document by painting them in the DOM's [top layer](https://fullscreen.spec.whatwg.org/#top-layer) retaining their paint order in the actual DOM.

# Design
This section describes the high level design for this feature with a tentative API:

## Outgoing Document
The transition sequence starts with the prepare phase when the browser receives a trigger for a transition. Consider the following snippet of code:

```js
<div shared-id="shared-elem"></div>

// MPA version.

> addEventListener("navigate", (event) => {

Reviewer: Does setData cause a transition to occur? This API example is hard to understand.

Author: Hopefully the new example makes this clearer.

```
top-layer
├───shared-old(root)
├───shared-container(#A)
├        ├───shared-old(#A)
├───shared-container(#B)
         ├───shared-old(#B)
```

* The top-layer state is copied to the incoming Document while its loading. The properties on the shared-container come from the cached [computed properties](#computed-properties) of the shared element in the outgoing Document. The following is an example of the incoming document when it is ready for first render.
```
HTML (root)
├───body
    ├───div "shared-id=A"
    ├───div "shared-id=C"
```

* After ready for first render, the top layer is populated with the Transition DOM for the incoming Document. The properties on the shared-container flip to the live [computed properties](#computed-properties) of the shared element in the incoming Document.
```

> top-layer

Reviewer: You could use this kind of diagram, or maybe just HTML?

<top-layer>
  <shared-new>
 ...

Author: HTML is good enough. Done.

Note : Supporting this requires the browser to copy the complete ComputedStyle for the pseudo elements in the Transition DOM from outgoing to incoming document, as opposed to only the [computed properties](#computed-properties).

```
.shared-element {
  border: 10px solid black;
  border-radius: 10% 10%;
  box-shadow: 0px 0px 10px;
  
  &:transition {
    /* Retain the border to ensure it is painted transparent but the box size is unchanged. */
    border: 10px solid transparent;
    border-radius: none;
    box-shadow: none;
  }
}

::shared-container(#shared-id) {
  border: 10px solid black;
  border-radius: 10% 10%;
  box-shadow: 0px 0px 10px;
}

::shared-old(#shared-id) {
  top: -10px;
  left: -10px;
}
```

> An alternate approach to this is to support this natively in the browser by introducing a new content-element() function. This function would behave similar to the element() function except skipping the following properties when painting the element: box decorations and visual effects which generate a stacking context. The image will also be sized to the element's content-box (as opposed to the border-box used by the element() function). The motivation for supporting this natively would be to ensure ease of use for developers.

Reviewer: You should clearly separate any alternatives into an "alternatives considered" section.

Author: Done.

This feature provides developers with the same capability on the web, irrespective of whether the scenes in a transition are rendered across Documents.

# Use-Cases
A visual demo of the transition patterns targeted by this feature are [here](https://material.io/design/motion/the-motion-system.html#transition-patterns). The following is a summary of the semantics of these transition patterns :

* Root Transitions : The full page content animates between 2 web pages with an optional static UI element on top. [Shared axis](https://material.io/design/motion/the-motion-system.html#shared-axis) shows an example.
* Shared Element Transitions : A persistent UI element morphs into another (which could be a UI element on the next page or the whole page) changing its shape and content. [Container transform](https://material.io/design/motion/the-motion-system.html#container-transform) shows an example.
* Entry/Exit Transitions : A UI element animates as it exits or enters the screen. This [issue](https://github.com/WICG/shared-element-transitions/issues/37) shows an example.

These transitions should be feasible in SPAs (Single Page Apps) and MPAs (Multi Page Apps).

# Design

> Let's take the example below which shows how the API can be used by a developer to animate the background and a shared element on a same origin navigation (MPA). The SPA equivalent of this case is one where the old document is mutated into the new document via DOM APIs. See [API Extensions](#api-extensions) for more code examples.

Reviewer: Add a note that this is a possible purely declarative API shape, but is mostly to be considered a placeholder in order to describe the steps?

Author: Sure, I added a note saying that the API shape is tentative. Let's avoid mentioning how much declarative vs imperative it needs to be.

### Old Document
```

> <html>

Reviewer: Add this to the document:

<a href="new-doc-url">Click me</a>

Author: This is taking me to a 404 error page. Wrong link?

Reviewer: I'm just suggesting you add an anchor link to the old page, pointing at the new page, so that it's a complete demo.

      top: 100px;
      width: 200px;
      height: 200px;
      background-color: green;
    }
    
    ::shared-container(#header-id) {
      animation: ::shared-container-header-id 1s ease-in;
    }
  </style>
</head>
<body>
  <div class="animated" sharedId="header-id" id="header">Shared Element</div>
</body>
</html>
```

The steps taken by the browser during the transition are as follows.

1. When a navigation is initiated on the old Document, create the following pseudo elements in the top layer[^1]. Note that a shared element must not be nested inside another shared element :

    a. A container and child replaced element for each element with the sharedId attribute. These are identified via ::shared-container(#sharedId) and ::shared-old(#sharedId) respectively.

    b. A replaced element for the root/html element identified via ::shared-old(#root).
    
    c. The box hierarchy in the top layer stacking context is :
    
    ```
    ├───shared-old(root)
    ├───shared-container(#header)
         ├───shared-old(#header)
    ```

2. Apply the following UA stylesheet to the pseudo elements on the old page :
```
// "root" is a reserved keyword for the html element.

> ::shared-old(#root), ::shared-container(#header-id) {

Reviewer: We won't be able to get away with reserved keywords here. I suggest renaming to ::shared-old-root.

Author: Done.

  position: fixed;
  top: 0px;
  left: 0px;
}

::shared-old(#root) {
  width: 100vw;
  height: 100vh;
  // The output of element() function on the root element.
  content: element(html);
}

::shared-container(#header-id) {
  // This size is is chosen exactly according to the #header-id element's
  // border box dimensions after layout.
  width: 100px;
  height: 100px;
  
  // A transform positioning the element relative to the viewport so that it overlaps
  // exactly with its screen coordinates and orientation it had on the old page.

>   transform: translate(0px, 308px);

Reviewer: ```suggestion
  transform: translate(8px, 308px);
```

Author: Done.

iii. In step 7 when applying a UA stylesheet to pseudo elements on the new page, the saved style is also applied to the old pseudo elements :
```
::shared-old(#root) {
  position: absolute;
  width: 100%;
  height: 100%;
  
  // This is the saved output referenced in step 3.
  content: cached-element(html);
  top: -10px;
  left: -10px;
}
```

## Modifications to element()
The following changes will be made to the element() spec as a part of this proposal. The element captured by this function is the target element :

* The target element must have paint containment (contain:paint) to ensure the element is the containing block for all positioned descendents and generates a stacking context.
* The target element must disallow fragmentation (break-inside:avoid).
* A new cached-element() function is introduced to refer to the saved output of the element() function in step 3) of the design.
* Nested shared elements are omitted from the output of element() function.
* The special cases when running the element() function on the html element are :
    * The natural size for the generated image is the visual viewport bounds.

>     * When creating the image, the element is drawn on a canvas with the background color of the document.

Reviewer: Also "shared elements captured via element() are not painted in the regular dom"

Author: Done.

# Introduction
When a user navigates on the web, they get to see the inner workings of web experiences: flash of white followed by a piece-meal rendering phase. This sequenced user experience results in a higher cognitive load because the user has to connect the dots between where they were, and where they are. Not only that, a smooth animation to transition between scenes also reduces the loading latency perceived by the user even if the actual loading time is the same. For these reasons, most platforms provide easy to use primitives to enable developers to build seamless transitions: [Android](https://developer.android.com/training/transitions/start-activity), [iOS/Mac](https://developer.apple.com/documentation/uikit/uimodaltransitionstyle) and [Windows](https://docs.microsoft.com/en-us/windows/apps/design/motion/page-transitions).

> This feature provides developers with the same capability on the web, irrespective of whether the scenes in a transition are rendered across Documents.

Reviewer: ```suggestion
Shared Element Transitions provides developers with the same capability on the web, irrespective of whether the scenes in a transition are rendered across Documents.
```

# Use-Cases

> A visual demo of the transition patterns targeted by this feature are [here](https://material.io/design/motion/the-motion-system.html#transition-patterns). The following is a summary of the semantics of these transition patterns :

Reviewer: ```suggestion
A visual demo of some example transition patterns targeted by this feature are [here](https://material.io/design/motion/the-motion-system.html#transition-patterns). The following is a summary of the semantics of these transition patterns :
```

> * Root Transitions : The full page content animates between 2 web pages with an optional static UI element on top. [Shared axis](https://material.io/design/motion/the-motion-system.html#shared-axis) shows an example.

Reviewer: ```suggestion
* Root Transitions : The full page content animates between two web pages with an optional static UI element on top. [Shared axis](https://material.io/design/motion/the-motion-system.html#shared-axis) shows an example.
```

* Entry/Exit Transitions : A UI element animates as it exits or enters the screen. This [issue](https://github.com/WICG/shared-element-transitions/issues/37) shows an example.

These transitions should be feasible in SPAs (Single Page Apps) and MPAs (Multi Page Apps).

# Design
Let's take the example below which shows how the API can be used by a developer to animate the background and a shared element on a same origin navigation (MPA). The SPA equivalent of this case is one where the old document is mutated into the new document via DOM APIs. See [API Extensions](#api-extensions) for more code examples.

Note that the API shape below is tentative and used to explain the core feature design.

### Old Document
```
<html>
<head>
  <style>
    body {
      background-color: blue;
    }
    .animated {
      position: relative;
      top: 300px;
      width: 100px;
      height: 100px;
      background-color: red;
    }
  </style>
</head>
<body>

>   <div class="animated" sharedId="header-id" id="header">Shared Element</div>

Reviewer: ```suggestion
  <div class="animated" sharedid="header-id" id="header">Shared Element</div>
```

  <a href="new-document.html">Click Me</a>
</body>
</html>
```

### New Document
```
<html>
<head>
  <style>
    body {
      background-color: lightblue;
    }
    .animated {
      position: relative;
      top: 100px;
      width: 200px;
      height: 200px;
      background-color: green;
    }
    

>     ::shared-container(#header-id) {

Reviewer: ```suggestion
    ::shared-container(header-id) {
```

      animation: ::shared-container-header-id 1s ease-in;
    }
  </style>
</head>
<body>

>   <div class="animated" sharedId="header-id" id="header">Shared Element</div>

Reviewer: ```suggestion
  <div class="animated" sharedid="header-id" id="header">Shared Element</div>
```

</body>
</html>
```

The steps taken by the browser during the transition are as follows.

1. When the user presses "Click Me" and a navigation is initiated on the old Document, create the following pseudo-elements in the top layer[^1]. Note that a shared element must not be nested inside another shared element :

>     a. A container pseudo-element and child replaced pseudo-element for each element with the sharedId attribute. These are identified via ::shared-container(#sharedId) and ::shared-old(#sharedId) respectively.

Reviewer: ```suggestion
    a. A container pseudo-element and child replaced pseudo-element for each element with the `sharedid` attribute. These are identified via ::shared-container(shared_id) and ::shared-old(shared_id) respectively,
     where shared_id is the value of the `sharedid` attribute.
```

    b. A replaced pseudo-element for the root/html element identified via ::shared-old-root.
    
    c. The box hierarchy in the top layer stacking context is :
    
    ```
    ├───shared-old-root
    ├───shared-container(#header)
         ├───shared-old(#header)
    ```

2. Apply the following UA stylesheet to the pseudo elements on the old page :
```
::shared-old-root, ::shared-container(#header-id) {
  position: fixed;
  top: 0px;
  left: 0px;
}

::shared-old-root {
  width: 100vw;
  height: 100vh;
  // The output of element() function on the root element.
  content: element(html);
}

> ::shared-container(#header-id) {

Reviewer: ```suggestion
::shared-container(header-id) {
```

  // This size is is chosen exactly according to the #header-id element's
  // border box dimensions after layout.
  width: 100px;
  height: 100px;
  
  // A transform positioning the element relative to the viewport so that it overlaps
  // exactly with its screen coordinates and orientation it had on the old page.
  transform: translate(8px, 308px);
}

> ::shared-old(#header-id) {

Reviewer: ```suggestion
::shared-old(header-id) {
```

  position: absolute;
  width: 100%;
  height: 100%;
  content: element(#header);
}
```

3. Save the output of the element() function for each pseudo element referenced above with the computed size and transform applied to container elements.

4. Navigate to the new page leaving the last rendered pixels of the old page on screen.

5. When the new page loads, suppress rendering until resources required for first render have been fetched.

6. Once the page is ready for first render, create the following pseudo elements in the top layer. The pseudo elements are kept in sync with the corresponding shared elements in the DOM until the transition completes as specified in step 8 :

    a. A container pseudo-element and child replaced pseudo-element for each shared element on the old page using state saved in step 3.

>     b. A container pseudo-element and child replaced pseudo-element for each element with sharedId attribute on the new page. The container is reused if already present.

Reviewer: ```suggestion
    b. A container pseudo-element and child replaced pseudo-element for each element with the `sharedid` attribute on the new page. The container is reused if already present.
```

    c. A replaced element for the root/html element identified via ::shared-new-root.

    d. The box hierarchy in the top layer stacking context is :
    
    ```
    ├───shared-new-root
    ├───shared-old-root
    ├───shared-container(#header)
         ├───shared-new(#header)
         ├───shared-old(#header)
    ```

7. Apply the following UA stylesheet to the pseudo elements on the new page.
```

> ::shared-old-root, ::shared-new-root, ::shared-container(#header-id) {

Reviewer: ```suggestion
::shared-old-root, ::shared-new-root, ::shared-container(header-id) {
```

  position: fixed;
  top: 0px;
  left: 0px;
}

::shared-old-root, ::shared-new-root {
  width: 100vw;
  height: 100vh;
  content: element(html);
}

// Update the container's size and transform to the shared element on the new page.
::shared-container(#header-id) {
  width: 200px;
  height: 200px;
  transform: translate(8px, 108px);
  
  // The blend mode referenced below is not currently exposed with mix-blend-mode.
  isolation: isolate;
  mix-blend-mode: [plus-lighter](https://drafts.fxtf.org/compositing/#porterduffcompositingoperators_plus_lighter);
}

> ::shared-old(#header-id), ::shared-new(#header-id) {

Reviewer: ```suggestion
::shared-old(#header-id), ::shared-new(header-id) {
```

  position: absolute;
  width: 100%;
  height: 100%;
}

::shared-old-root {
  // This is the saved output referenced in step 3.
  content: cached-element(html);
}
::shared-old(#header-id) {
  content: cached-element(#header);
}

::shared-new-root {
  content: element(html);
}
::shared-new(#header-id) {
  content: element(#header);
}

// Default animations added by the UA which can be overridden by the developer stylesheet/script.
@keyframes ::shared-new-fade-in {
  0% {opacity: 0;}
  100% {opacity: 1;}
}

> ::shared-new-root, ::shared-new(#header-id) {

Reviewer: ```suggestion
::shared-new-root, ::shared-new(header-id) {
```

  animation: ::shared-new-fade-in 0.25s;
}

@keyframes ::shared-old-fade-out {
  0% {opacity: 1;}
  100% {opacity: 0;}
}

> ::shared-old-root, ::shared-old(#header-id) {

Reviewer: ```suggestion
::shared-old-root, ::shared-old(header-id) {
```

  animation: ::shared-old-fade-out 0.25s;
}

> // Generated for each shared element with the syntax shared-container-sharedId.

Reviewer: ```suggestion
// Generated for each shared element with the syntax shared-container-shared_d.
```

Author: Fixed.

@keyframes ::shared-container-header-id {
  from {
    width: 100px;
    height: 100px;
    transform: translate(0px, 308px);
  }
}

> ::shared-container(#header-id) {

Reviewer: ```suggestion
::shared-container(header-id) {
```

## Animating Box Decoration CSS Properties
A common capability desirable during transitions is to interpolate styles like border-radius that change the element's shape. The [container transform](https://material.io/design/motion/the-motion-system.html#container-transform) examples show a visual demo of that. Painting properties like the element's border within its image when using the element() function makes this difficult.

Consider the same example[^2] as above with the addition of box decorations to the shared element.

### Old Document
```
<html>
<head>
  <style>
    body {
      background-color: blue;
    }
    
    .animated {
      position: relative;
      top: 300px;
      width: 100px;
      height: 100px;
      background-color: red;
      
      border: 10px solid black;
      border-radius: 10% 10%;
      box-shadow: 0px 0px 10px;
      &:transition {
        /* Retain the border to ensure it is painted transparent but the box size is unchanged. */
        border: 10px solid transparent;
        border-radius: none;
        box-shadow: none;
      }
    }
    

>     ::shared-container(#header-id) {

Reviewer: ```suggestion
    ::shared-container(header-id) {
```

      border: 10px solid black;
      border-radius: 10% 10%;
      box-shadow: 0px 0px 10px;
    }
    
    // Offset the image to account for the border in the snapshot.

>     ::shared-old(#header-id) {

Reviewer: ```suggestion
    ::shared-old(header-id) {
```

      top: -10px;
      left: -10px;
    }
  </style>
</head>
<body>

>   <div class="animated" sharedId="header-id" id="header">Shared Element</div>

Reviewer: ```suggestion
  <div class="animated" sharedid="header-id" id="header">Shared Element</div>
```

</body>
</html>
```

### New Document
```
<html>
<head>
  <style>
    body {
      background-color: blue;
    }
    
    .animated {
      position: relative;
      top: 100px;
      width: 200px;
      height: 200px;
      background-color: green;
      
      border: 5px solid black;
      border-radius: 50% 50%;
      box-shadow: 0px 0px 5px;
      
      &:transition {
        border: 5px solid transparent;
        border-radius: none;
        box-shadow: none;
      }
    }
    

>     ::shared-container(#header-id) {

Reviewer: ```suggestion
    ::shared-container(header-id) {
```

        border: 5px solid black;
        border-radius: 50% 50%;
        box-shadow: 0px 0px 5px;
    }
    

>     ::shared-new(#header-id) {

Reviewer: ```suggestion
    ::shared-new(header-id) {
```

      top: -5px;
      left: -5px;
    }
    
    @keyframes border-animation {
      from {
        border: 10px solid black;
        border-radius: 10% 10%;
        box-shadow: 0px 0px 10px;
      }
    }

>     ::shared-container(#header-id) {

Reviewer: ```suggestion
    ::shared-container(header-id) {
```

      animation: ::shared-container-header-id 1s ease-in, border-animation 1s ease-in;
    }
    
    @keyframes position-animation {
      from {
        top: -10px;
        left: -10px;
      }
      to {
        top: -5px;
        left: -5px;
      }
    }

>     ::shared-old(#header-id), ::shared-new(#header-id) {

Reviewer: ```suggestion
    ::shared-old(header-id), ::shared-new(header-id) {
```

      animation: position-animation 1s ease-in;
    }
    
  </style>
</head>
<body>

>   <div class="animated" sharedId="header-id" id="header">Shared Element</div>

Reviewer: ```suggestion
  <div class="animated" sharedid="header-id" id="header">Shared Element</div>
```

</body>
</html>
```

The additional steps taken by the browser in the example above are :

i. When a pseudo element is created for shared elements in the old page in step 1, the pseudo-class "transition" is enabled for each shared element.

ii. In step 3 when saving state from the old page, the completed ComputedStyle on the pseudo elements is saved in addition to the computed size and transform.

iii. In step 7 when applying a UA stylesheet to pseudo elements on the new page, the saved style is also applied to the old pseudo elements :
```
::shared-old-root {
  position: absolute;
  width: 100%;
  height: 100%;
  
  // This is the saved output referenced in step 3.
  content: cached-element(html);
  top: -10px;
  left: -10px;
}
```

## Modifications to element()
The following changes will be made to the element() spec as a part of this proposal. The element captured by this function is the target element :

* The target element must have paint containment (contain:paint) to ensure the element is the containing block for all positioned descendents and generates a stacking context.

> * The target element must disallow fragmentation (break-inside:avoid).

Reviewer: ```suggestion
* The target element must disallow fragmentation (similar to `break-inside:avoid`).
```

> * A new cached-element() function is introduced to refer to the saved output of the element() function in step 3 of the design.

Reviewer: ```suggestion
* A new cached-element() function is introduced to refer to the saved output of the element() function in step 3 of the design. (This part may not be developer exposed though.)
```

* Nested shared elements are omitted from the output of element() function.

> * Elements captured using element() and displayed via pseudo-elements during the transition are not painted in the regular DOM.

Reviewer: ```suggestion
* Elements captured using element() and displayed via pseudo-elements during the transition are not painted in the regular DOM - they behave as if they have `content-visibility: hidden`, except that they don't have `contain:size`.
```

    * The natural size for the generated image is the visual viewport bounds.
    * When creating the image, the element is drawn on a canvas with the background color of the document.

# API Extensions
## SPA
The SPA code requires the addition of script APIs which provide the equivalent of "navigation" and "ready for first render" events referenced in step 1 and 4 of the design. The rest of the code is identical between MPA and SPA.

```js
function handleTransition() {
  document.documentTransition.prepare(async () => {
    await loadNextPage();
  });
}
```

* The prepare API initiates step 1 to 3 to save the state of shared elements in the current DOM. The API takes a callback once the save operation finishes executing.
* The async callback initiates load of the next page and initiates step 4 which suppresses rendering.
* When the callback returns, the new DOM is considered ready for first render. This starts step 5 onwards to create new pseudo elements and start animations.

## Additional Script APIs
The following example shows how developers can configure the transition in script for an MPA.

### Old Document
```js
addEventListener("navigate", (event) => {

>   // Add sharedId attribute to elements to offer for the transition

Reviewer: ```suggestion
  // Add sharedid attribute to elements to offer for the transition
```

  // based on current document state.

>   document.querySelector(".header").sharedId="header-id";

Reviewer: ```suggestion
  document.querySelector(".header").sharedid="header-id";
```

  
  // setData can be used to pass opaque contextual information to the
  // new page. The argument type is |any|.
  document.documentTransition.setData({ version: 123 });
});
```

### New Document
```js
requestAnimationFrame(() => {
  let pendingTransition = document.documentTransition.getPendingTransition();
  if (pendingTransition.getData().version !== 123)
    return;
  
  // |offeredTransitionElements| provides a list of objects to access state
  // saved from the old page.
  let oldHeader = pendingTransition.offeredTransitionItems.get("header-id");
  if (oldHeader) {

>     // Add sharedId attribute to elements animated in the new DOM.

Reviewer: ```suggestion
    // Add sharedid attribute to elements animated in the new DOM.
```

>     document.querySelector(".header").sharedId="header-id";

Reviewer: ```suggestion
    document.querySelector(".header").sharedid="header-id";
```

    
    // Query the style information saved from the old page.
    let oldHeaderStyle = oldHeader.getContainerComputedStyle();
    
    // The pseudo elements for each shared element are associated with the root element.
    // The existing [pseudoElement](https://drafts.csswg.org/web-animations-1/#dom-keyframeeffect-pseudoelement) option can be used
    // to target them with Web Animations API.
    document.documentElement.animate(
      [{ width: oldHeaderStyle.width,
        height: oldHeaderStyle.height,
        transform: oldHeaderStyle.transform }],
      { duration: 1000,
        pseudoElement: '::shared-container(#header-id)' });
  }
});
```

# Alternatives Considered
## Heirarchical Properties
This proposal disallows a shared element to be nested inside another shared element. The restriction avoids the need to preserve the hierarchy of the shared elements and associated properties (transform, clip, effects inherited by descendents) when creating pseudo elements. This is a consideration for future iterations of the feature.

## Container/Child Split

> One consideration was to render each shared element using a replaced element directly instead of creating a container element. The motivation behind this split is to provide a stacking context to cross-fade the content of old and new shared elements. This was necessary to ensure blending identical pixels is a no-op using [plus-lighter](https://drafts.fxtf.org/compositing/#porterduffcompositingoperators_plus_lighter) blending. While same-origin transitions could work around this, it enables future extensibility for cross-origin transitions where cross-fading identical images would be common.

Reviewer: ```suggestion
One consideration is to render each shared element using a replaced element directly instead of creating a container element. The motivation behind this split is to provide a stacking context to cross-fade the content of old and new shared elements. This was necessary to ensure blending identical pixels is a no-op using [plus-lighter](https://drafts.fxtf.org/compositing/#porterduffcompositingoperators_plus_lighter) blending. While same-origin transitions could work around this, it enables future extensibility for cross-origin transitions where cross-fading identical images would be common.
```

## Natively Supporting Animating Box Decoration CSS Properties

> An alternate approach to the setup described in [Animating Box Decoration CSS Properties](#live-animatable-properties) is to support this natively in the browser by introducing a new content-element() function. This function would behave similar to the element() function except skipping the following properties when painting the element: box decorations and visual effects which generate a stacking context. The image will also be sized to the element's content-box (as opposed to the border-box used by the element() function). The motivation for supporting this natively would be to make these properties animatable instead of requiring developer side changes.

Reviewer: ```suggestion
An alternate approach to the setup described in [Animating Box Decoration CSS Properties](#live-animatable-properties) is to support this natively in the browser by introducing a new `content-element()` function. This function would behave similarl to the `element()` function except skipping the following properties when painting the element: box decorations and visual effects which generate a stacking context. The image will also be sized to the element's content-box (as opposed to the border-box used by the element() function). The motivation for supporting this natively would be to make these properties animatable instead of requiring developer side changes.
```

# Security/Privacy Considerations

> The security considerations below are limited to same-origin transitions :

Author: Clarified that they apply to cross-origin as well. But are a subset of the considerations for cross-origin.

# Introduction

When a user navigates on the web, they get to see the inner workings of web
experiences: flash of white followed by a piece-meal rendering phase. This
sequenced user experience results in a higher cognitive load because the user
has to connect the dots between where they were, and where they are. Not only
that, a smooth animation to transition between scenes also reduces the loading
latency perceived by the user even if the actual loading time is the same. For
these reasons, most platforms provide easy to use primitives to enable
developers to build seamless transitions:
[Android](https://developer.android.com/training/transitions/start-activity),
[iOS/Mac](https://developer.apple.com/documentation/uikit/uimodaltransitionstyle)
and
[Windows](https://docs.microsoft.com/en-us/windows/apps/design/motion/page-transitions).

Shared Element Transitions provides developers with the same capability on the
web, irrespective of whether the transitions are cross-document or
intra-document (SPA).

# Use-Cases

> A visual demo of some example transition patterns targeted by this feature are

Reviewer 2: typo? :)

> A visual demo of some example transition patterns targeted by this feature are

Author: Typo indeed. I'll resolve this.
